### PR TITLE
test: fix gh-5430 test

### LINKED ISSUE #9617: Strange `replication/gh-5430-cluster-mvcc` test execution.
Found in `3.1.0-entrypoint-35-gf58bfc97c`.

The test execution is strange.

1. Test takes more then 1 min to finish while looks like it should executed quite quickly looking at the test code.

I added logs to investigate the test timings.

<details>
<summary>test diff</summary>

```diff
--- a/test/replication/gh-5430-cluster-mvcc.test.lua
+++ b/test/replication/gh-5430-cluster-mvcc.test.lua
@@ -31,32 +31,44 @@ test_run:cmd('create server replica2 with '..
              'script="replication/gh-5430-mvcc-replica2.lua"')
 test_run:cmd('start server replica2')

 test_run:switch('master')
+log = require('log')
 box.error.injection.set('ERRINJ_WAL_DELAY', true)
+log.info('delay set finish')

 test_run:switch('replica1')
 _ = require('fiber').create(function() box.cfg{replication_anon = false} end)
 str = string.format('registering replica %s', box.info.uuid):gsub('-', '%%-')
 _ = test_run:wait_log('master', str)
+test_run:switch('master')
+log.info('log1 wait finish')

 test_run:switch('replica2')
 _ = require('fiber').create(function() box.cfg{replication_anon = false} end)
 str = string.format('registering replica %s', box.info.uuid):gsub('-', '%%-')
 _ = test_run:wait_log('master', str)
+test_run:switch('master')
+log.info('log2 wait finish')

 test_run:switch('master')
 box.error.injection.set('ERRINJ_WAL_DELAY', false)
+log.info('delay unset finish')

 test_run:switch('replica1')
 test_run:wait_cond(function() return box.info.id > 1 end)
+test_run:switch('master')
+log.info('id1 wait finish')

 test_run:switch('replica2')
 test_run:wait_cond(function() return box.info.id > 1 end)
+test_run:switch('master')
+log.info('id2 wait finish')
```

</details>

<details>
<summary>logs</summary>

```
2024-01-24 15:22:44.016 [488658] main/123/console/unix/:/tarantool I> delay set finish
2024-01-24 15:22:44.022 [488658] relay/unix/:(socket)/101/main coio.c:352 E> SocketError: unexpected EOF when reading from socket, called on fd 29, aka unix/:(socket), peer of unix/:(socket): Broken pipe
2024-01-24 15:22:44.022 [488658] relay/unix/:(socket)/101/main I> exiting the relay loop
2024-01-24 15:22:44.123 [488658] main/111/main I> registering replica 8d803249-9e15-4aba-9252-8228cbc05752 at fd 31, aka unix/:(socket), peer of unix/:(socket)
2024-01-24 15:22:44.123 [488658] main/111/main I> assigned id 2 to replica 8d803249-9e15-4aba-9252-8228cbc05752
2024-01-24 15:22:44.123 [488658] main/111/main I> update replication_synchro_quorum = 2
2024-01-24 15:22:44.123 [488658] main/111/main I> RAFT: fencing paused
2024-01-24 15:22:44.127 [488658] main/169/console/unix/:/tarantool I> log1 wait finish
2024-01-24 15:23:44.142 [488658] main/52733/console/unix/:/tarantool I> log2 wait finish
2024-01-24 15:23:44.156 [488658] main txn.c:826 W> too long WAL write: 1 rows at LSN 2: 60.033 sec
2024-01-24 15:23:44.157 [488658] main/52741/console/unix/:/tarantool I> delay unset finish
```

</details>

So looks like instead of successful wait of event of registering second replica we simply have wait timeout (note 1 min delay in logs).

2. In the scope of PR https://github.com/tarantool/tarantool/pull/9604 I'm going to add next hunk to console:

```diff
@@ -985,6 +986,9 @@ local function client_handler(client, _peer)
     state:print(string.format("%-63s\n%-63s\n",
         "Tarantool ".. version.." (Lua console)",
         "type 'help' for interactive help"))
+    box.ctl.on_shutdown(function()
+        state.running = false
+    end)
     repl(state)
     session_internal.run_on_disconnect()
 end
```
Somehow after this change I got issues with stopping `master` server. 
```
[001] * [QA Notice]
[001] *
[001] * The server 'gh-5430-mvcc-master' does not stop during 5 seconds after the 15 (SIGTERM) signal.
[001] * Info: process 488354 [R (running); tarantool gh-5430-mvcc-master.lua <running>]
[001] * Sending SIGKILL...
```

### PR DESCRIPTION
**Pull request by CuriousGeorgiy**
This patch fixes the gh-5430 test.

Closes #9617

### CODE REVIEW

File: test/replication/gh-5430-cluster-mvcc.test.lua
Comment (Gerold103) [inline test/replication/gh-5430-cluster-mvcc.test.lua:52]: The test was supposed to get 2 concurrent `_cluster` txns running and then unblock WAL. But now the test might be starting the registration of `replica1` on the master, then finishing the WAL delay, and then starting the registration of `replica2`. Because you start the latter in a fiber without waiting for anything. It could as well not reach the master until you make the second `wait_log` after the first replica is already registered.

Why was `wait_log` timing out?

**Review by Gerold103 — COMMENTED** [review]
Thanks for the patch!

File: test/replication/gh-5430-cluster-mvcc.test.lua
Comment (CuriousGeorgiy) [reply test/replication/gh-5430-cluster-mvcc.test.lua:52]: Sorry, I misunderstood the purpose of the test.

It seems like after 70a68836 we are unable to register an anonymous replica while its relay is not finished:
https://github.com/tarantool/tarantool/blob/9e616ab6f2f94ba2bcad3044187c474a1c05a8c5/src/box/box.cc#L4422-L4427

At the same time, the anonymous replica's relay cannot finish, because it is blocked by `cbus_unpair` in `wal_clear_watcher`:
https://github.com/tarantool/tarantool/blob/9e616ab6f2f94ba2bcad3044187c474a1c05a8c5/src/box/relay.cc#L1068-L1074

https://github.com/tarantool/tarantool/blob/9e616ab6f2f94ba2bcad3044187c474a1c05a8c5/src/lib/core/cbus.h#L440-L442

Comment (Gerold103) [reply test/replication/gh-5430-cluster-mvcc.test.lua:52]: Hm. The case is definitely still valid, but the test indeed needs a rework. Since this is MVCC, I would carefully assume, that instead of box.error.injection.set('ERRINJ_WAL_DELAY', true) we could set an on-replace trigger for _cluster, which would yield until allowed to continue. That way we could get 2 txns on _cluster in parallel.

If this won't work somehow, I suggest to move ERRINJ_REPLICA_JOIN_DELAY before box_register_replica(). This errinj is not used right now for anything, so we can reuse it here instead of ERRINJ_WAL_DELAY. Then WAL isn't blocked, cbus_unpair with WAL endpoints will work, and yet the replicas won't register in turns.

Once we see that both of them are registering replica in the logs, we can set ERRINJ_WAL_DELAY and unset ERRINJ_REPLICA_JOIN_DELAY.

Comment (CuriousGeorgiy) [reply test/replication/gh-5430-cluster-mvcc.test.lua]: Thanks for the help!

We cannot do it the first way because it is forbidden to yield during system space changes.

But the second way seems to be okay, please check the test changes.

File: test/replication/gh-5430-cluster-mvcc.test.lua
Comment (Gerold103) [inline test/replication/gh-5430-cluster-mvcc.test.lua:44]: Do you need this new switch here? You are already on `replica2` in this place.

File: test/replication/gh-5430-cluster-mvcc.test.lua
Comment (Gerold103) [inline test/replication/gh-5430-cluster-mvcc.test.lua:55]: Nit: in those old diff tests you can use `/` as a newline escape symbol. It would make this 2 lines shorter. Up to you of course.

**Review by Gerold103 — COMMENTED** [review]
Great, that looks nice and neat! Thanks for trying to save the test!

File: test/replication/gh-5430-cluster-mvcc.test.lua
Comment (CuriousGeorgiy) [reply test/replication/gh-5430-cluster-mvcc.test.lua]: Sorry, it's a typo, dropped.

File: test/replication/gh-5430-cluster-mvcc.test.lua
Comment (CuriousGeorgiy) [reply test/replication/gh-5430-cluster-mvcc.test.lua:54]: I prefer this pattern for defining functions in diff tests.

**Review by Gerold103 — APPROVED** [review]

**Pull request merged**
